### PR TITLE
Error in wasm-ld when essential symbols are missing. NFC

### PR DESCRIPTION
**Pull request by sbc100**
Create new  `REQUIRED_SYMBOLS` internal setting.  Any symbols in
this list are exported with `--export` rather than `--export-if-defined`
which means if they are missing the link will fail during `wasm-ld`.

Like `EXPORT_IF_DEFINED` this list contains native symbol names without
any `_` prefixes.

These are internal settings so this change should not be user-visible.
This allows for a simpler/shorter `wasm-ld` commands and earlier failure
when one of these required symbols in missing.

This is useful for #15703.

### CODE REVIEW

File: tools/toolchain_profiler.py [hunk 1]
```diff
@@ -26,10 +26,13 @@ def __init__(self, name):
...
 
-  def __exit__(self, type, value, traceback):
+  def __exit__(self, exc_type, value, traceback):
     # When a block ends debug log the total duration.
     now = time.time()
-    logger.debug('block "%s" took %.2f seconds', self.name, now - self.start)
+    if exc_type:
+      logger.debug('block "%s" raised an exception after %.2f seconds', self.name, now - self.start)
+    else:
+      logger.debug('block "%s" took %.2f seconds', self.name, now - self.start)
```
Comment (kripken) [inline tools/toolchain_profiler.py:35]: unrelated?

File: tools/deps_info.py [hunk 2]
```diff
@@ -141,7 +141,6 @@
   'emscripten_set_visibilitychange_callback_on_thread': ['malloc', 'free'],
   'emscripten_set_wheel_callback_on_thread': ['malloc', 'free'],
   'emscripten_webgl_create_context': ['malloc'],
-  'emscripten_webgl_destroy_context': ['emscripten_webgl_make_context_current', 'emscripten_webgl_get_current_context'],
```
Comment (kripken) [inline tools/deps_info.py:144]: unrelated?

File: tests/test_core.py [hunk 3]
```diff
@@ -4249,7 +4249,7 @@ def test_dylink_postsets_chunking(self):
     'libcxx': ('libc,libc++,libmalloc,libc++abi',),
     'all': ('1',),
     'missing': ('libc,libmalloc', False, False, False),
-    'missing_assertions': ('libc,libmalloc', False, False, True),
+    'missing_assertions': ('libc,libc++abi,libmalloc', False, False, True),
```
Comment (kripken) [inline tests/test_core.py:4252]: why does this change?

File: emcc.py [hunk 4]
```diff
@@ -1691,13 +1691,12 @@ def default_setting(name, new_default):
...
-    settings.EXPORTED_FUNCTIONS += [
-        # This needs to be exported on the Module object too so it's visible
-        # to side modules too.
-        '___heap_base',
-        # Unconditional dependency in library_dylink.js
-        '_setThrew',
-    ]
+    # This needs to be exported on the Module object too so it's visible
+    # to side modules too.
+    settings.EXPORTED_FUNCTIONS.append('___heap_base')
```
Comment (kripken) [inline emcc.py:1696]: why does this not use REQUIRED_SYMBOLS as well?

**Review by kripken — COMMENTED** [review]
Nice idea! I like using a simpler internal setting after handling the more complicated user-facing one. And this can be a good step towards eventually removing the prefixing entirely in #8380

File: tools/toolchain_profiler.py (hunk 1 above)
Comment (sbc100) [reply tools/toolchain_profiler.py:35]: Yes.  This just makes debug logging more clear in case of a crash.   Want me to split it out?

Comment (kripken) [reply tools/toolchain_profiler.py:35]: Seems ok to leave it in to me. Just wanted to check it was intentional.

File: tools/deps_info.py (hunk 2 above)
Comment (sbc100) [reply tools/deps_info.py:144]: This change is needed for this PR, but it could be split out and landed first if you like.

The reason is this:

Neither `emscripten_webgl_make_context_current` nor  `emscripten_webgl_get_current_context` are defined in native code.  They are both defined in JS:

https://github.com/emscripten-core/emscripten/blob/8c34d43d4407463aba9e196a55af2bf392692253/src/library_html5_webgl.js#L223

https://github.com/emscripten-core/emscripten/blob/8c34d43d4407463aba9e196a55af2bf392692253/src/library_html5_webgl.js#L57


What is more the dependency between `emscripten_webgl_destroy_context` and these two functions is expressed already in JS:

https://github.com/emscripten-core/emscripten/blob/8c34d43d4407463aba9e196a55af2bf392692253/src/library_html5_webgl.js#L313

Dependencies between JS functions and other JS functions don't make any sense in deps_info.py so I'm not sure why this existed.

Prior to this PR such dependencies were harmless but went unnoticed..  with this PR these get turned into required native exports which will cause wasm-ld to fail (quite rightly) with "undefined symbol".

Do you want me to split this out even though it is needed for to this change? 


Having this unneeded

Comment (sbc100) [reply tools/deps_info.py:144]: Split out into https://github.com/emscripten-core/emscripten/pull/15719

File: emcc.py (hunk 4 above)
Comment (sbc100) [reply emcc.py]: When `RELOCATABLE` code `__heap_base` (like `__stack_pointer`) is defined, in JS code, not in native code, so it can't go in the `REQUIRED_SYMBOLS`.

File: tests/test_core.py (hunk 3 above)
Comment (sbc100) [reply tests/test_core.py:4252]: Reverted..

**Review by kripken — APPROVED** [review]

**Comment by Birch-san** [comment]
I think this breaks `-nostdlib`.

I'm trying to demonstrate a minimal program `vector.cpp`:

```c++
#include <new>
struct Vec2 {
  float x, y;
  Vec2(float x, float y) : x(x), y(y) {};
};

extern "C" {
  void __attribute__((used)) Vec2_Vec2_2(Vec2* ptr, float x, float y) {
    new(ptr) Vec2(x, y);
  }
}
```

Building with:

```c++
em++ -Wl,--no-entry -nostdlib -o vector.wasm vector.cpp
```

Build worked fine in 2.0.27, but now I've updated to latest (3.1.6) it's failing with:

```
wasm-ld: error: symbol exported via --export not found: stackSave
wasm-ld: error: symbol exported via --export not found: stackRestore
wasm-ld: error: symbol exported via --export not found: stackAlloc
wasm-ld: error: symbol exported via --export not found: __errno_location
em++: error: '/Users/birch/git/emsdk/upstream/bin/wasm-ld -o vector.wasm --no-entry /var/folders/hn/_fwb1slj4wv_gg_k39xyn_g80000gn/T/emscripten_temp_vlqc10wx/vector_0.o -L/Users/birch/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --import-undefined --strip-debug --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export-table -z stack-size=5242880 --initial-memory=16777216 --max-memory=16777216 --global-base=1024' failed (returned 1)
```

Build works in 3.1.6 if I remove `-nostdlib`.

**Comment by sbc100** [comment]
Yeah, its not clear how `-nostdlib` should work in this case.   Those symbol are required by the emscripten runtime, so if you don't include them via the standard library you still need to supply them somehow, otheriwse your program won't work.

Just like if have have a program that calles `printf` and you link with `-nostdlib` then if up to you to supply a definition of `printf`.    In this case its not clear to me if the program does actually depend on `__errno_location` and the stack allocation functions.   Its quite likely that it does.   Does the resulting program run when you build with `ERROR_ON_UNDEFINED_SYMBOLS=0`?   Do you see references to the missing symbols in the generated JS (I would assume that answer is yes).
